Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -14,173 +14,222 @@ import Data.Semigroup
 #endif
 import qualified Data.List as L
 import qualified Data.IntMap as I
+import qualified Data.IntSet as S
 import qualified Data.Vector as V
 import qualified Data.Vector.Sized as SV
 import qualified Data.Vector.Mutable as MV
 import qualified Data.Vector.Unboxed as U
 import qualified Data.Vector.Unboxed.Sized as SU
-import qualified Math.NumberTheory.Primes.IntSet as PS
-import Math.NumberTheory.Roots
+import qualified Data.Mod as M
+import qualified Data.Mod.Word as MW
 import Math.NumberTheory.Primes
+import Math.NumberTheory.Logarithms
 import Math.NumberTheory.Moduli.Sqrt
 import Math.NumberTheory.Utils.FromIntegral
 import Math.NumberTheory.Primes.Factorisation.LinearAlgebra
+import Math.NumberTheory.Primes.Factorisation.TrialDivision
 import Control.Monad
 import Control.Monad.ST
 import GHC.TypeNats
 import Data.Proxy
 import Data.Kind
+import Data.Foldable
 import Data.Maybe
 import Data.Bit
 import Data.Bifunctor
+import Debug.Trace
 
--- This datatype stores factorisations of numbers. Its @sign@ describes the
--- sign of the number (True = Negative and False = Non-negative) and @primeSet@
--- stores the prime numbers in its factorisation.
-data SignedPrimeIntSet = SignedPrimeIntSet
-  { sign :: !Bool
-  , primeSet :: !PS.PrimeIntSet
-  } deriving (Show, Eq, Ord)
-
-insert :: Prime Int -> SignedPrimeIntSet -> SignedPrimeIntSet
-insert prime (SignedPrimeIntSet s ps) = SignedPrimeIntSet s (prime `PS.insert` ps)
-
--- | Given an odd positive composite Integer @n@ and Int parameters @b@ and @t@,
--- the Quadratic Sieve attempts to output @factor@, a factor of @n@. The
+import Math.NumberTheory.Roots
+-- | Given an odd positive composite Integer @n@ and Int parameters @b@, @t@ and
+-- @k@, the Quadratic Sieve attempts to output @factor@, a factor of @n@. The
 -- parameter @b@ controls the size of the factor base. This consists of all
 -- the relevant primes which are less than or equal to @b@. The parameter @t@
--- controls the length of the sieving interval.
-quadraticSieve :: Integer -> Int -> Int -> Integer
-quadraticSieve n b t = findFactor n $ findSquares n b t
+-- controls the length of the sieving interval. The parameter @k@ controls the
+-- maximum number of blocks for the sieve to go through (@2 ^ (k - 1)@).
+quadraticSieve :: Integer -> Int -> Int -> Int -> Integer
```

Comment:
Having three `Int`s is error-prone: one can easily swap them. Let's introduce
```haskell
data QuadraticSieveConfig = QuadraticSieveConfig 
  { qscFactorBase :: Int 
  , qscSievingInterval :: Int 
  , qscNumberOfBlocks :: Int 
  }
```
(you can come up with better names, these are just for example).